pgjdbc/src/test/java/org/postgresql/test/jdbc3/ParameterMetaDataTest.java

```diff
@@ -129,4 +140,75 @@ public void testTypeChangeWithUnknown() throws SQLException {
     rs.close();
   }
 
+  @Test
+  public void testGetParameterMetadata_cached() throws SQLException, IllegalAccessException {
+    // given
+    PreparedStatement pstmt =
+        con.prepareStatement("SELECT a FROM parametertest WHERE c = ? AND e = ?");
+
+    pstmt.setString(1, "anything");
+    pstmt.setTimestamp(2, new Timestamp(0L));
+
+    int expectedAmountOfExecutions =
+        binaryMode == BinaryMode.FORCE
+            ? 1
+            : Integer.parseInt(PGProperty.PREPARE_THRESHOLD.getDefaultValue());
+
+    for (int i = 0; i < expectedAmountOfExecutions; i++) {
+      pstmt.executeQuery();
+    }
+
+    // and
+    setUpSpyOnQueryExecutor();
+    QueryExecutor spyOnQueryExecutor = con.unwrap(PgConnection.class).getQueryExecutor();
+
+    // when
+    ParameterMetaData firstCall = pstmt.getParameterMetaData();
+    ParameterMetaData secondCall = pstmt.getParameterMetaData();
+    ParameterMetaData thirdCall = pstmt.getParameterMetaData();
+
+    // then
+    Mockito
+        .verify(spyOnQueryExecutor, Mockito.never())
+        .execute(
+            Mockito.any(Query.class),
+            Mockito.any(ParameterList.class),
+            Mockito.any(ResultHandler.class),
+            Mockito.anyInt(),
+            Mockito.anyInt(),
+            Mockito.anyInt()
+        );
+
+    Mockito
+        .verify(spyOnQueryExecutor, Mockito.times(3))
+        .requiresDescribe(
+            Mockito.any(Query.class),
+            Mockito.any(ParameterList.class)
+        );
+  }
+
+  private QueryExecutor setUpSpyOnQueryExecutor() throws SQLException, IllegalAccessException {
+    PgConnection pgConnection = con.unwrap(PgConnection.class);
+
+    Field queryExecutorField = ReflectionUtils.findFields(
+        PgConnection.class,
+        field -> field.getName().equals("queryExecutor"),
+        HierarchyTraversalMode.TOP_DOWN
+    ).get(0);
+
+    queryExecutorField.setAccessible(true);
+    QueryExecutor originalQueryExecutor = (QueryExecutor) queryExecutorField.get(pgConnection);
+
+    // this check is necessary to not spy on spy.
+    // It is required since the connection is shared between the tests, and so is QueryExecutor bind to it.
+    if (!Mockito.mockingDetails(originalQueryExecutor).isSpy()) {
+      QueryExecutor spyOnQueryExecutor = Mockito.spy(originalQueryExecutor);
+      queryExecutorField.set(pgConnection, spyOnQueryExecutor);
+      return spyOnQueryExecutor;
+    } else {
+      Mockito.clearInvocations(originalQueryExecutor);
+    }
```

Comment:
I would suggest placing shorter `if-else` branches first to make the code easier to read.

This case begs for flipping `if-else` as the `if` condition is negated.